io/slow5/slow5.go

```diff
@@ -0,0 +1,289 @@
+				numReadGroupsUint, err := strconv.ParseUint(values[1], 10, 32)
+				if err != nil {
+					return []Header{}, err
+				}
+				numReadGroups = uint32(numReadGroupsUint)
+				for id := uint32(0); id < numReadGroups; id++ {
+					emptyMap := make(map[string]string)
+					headers = append(headers, Header{Slow5Version: slow5Version, ReadGroupId: id, Attributes: emptyMap})
+				}
+			}
+		} else {
+			// Terminate if we hit the beginning of the raw read headers
+			if values[0] == "#char*" || values[0] == "#read_id" {
+				return headers, nil
+			}
+			// Check to make sure we have the right amount of information for the num_read_groups
+			if len(values) != int(numReadGroups+1) {
+				return []Header{}, fmt.Errorf("Improper amount of information for read groups. Needed %d, got %d, in line: %s", numReadGroups+1, len(values), line)
+			}
+			for id := 0; id < int(numReadGroups); id++ {
+				headers[id].Attributes[values[0]] = values[id+1]
+			}
+		}
+		lineNum++
+	}
+	return headers, errors.New("ReadGroup header never terminated. Improper file?")
+}
+
+var knownEndReasons = map[string]bool{"unknown": true,
+	"partial":                         true,
+	"mux_change":                      true,
+	"unblock_mux_change":              true,
+	"data_service_unblock_mux_change": true,
+	"signal_positive":                 true,
+	"signal_negative":                 true,
+}
+
+// ParseReads parses all reads in a slow5 file.
+func ParseReads(r io.Reader, reads chan<- Read, errorsChan chan<- error) {
+	headerMap := make(map[int]string)
+	endReasonMap := make(map[int]string)
+	scanner := bufio.NewScanner(r)
+	start := false
+	lineNum := 0
+	for scanner.Scan() {
+		line := scanner.Text()
+		values := strings.Split(line, "\t")
+		if !start {
+			// Get endReasonEnums. This is simply a string between enum{} that is used for the reasons that a read could have ended.
+			if values[0] == "#char*" {
+				for _, typeInfo := range values {
+					if strings.Contains(typeInfo, "enum") {
+						endReasonEnumsMinusPrefix := strings.TrimPrefix(typeInfo, "enum{")
+						endReasonEnumsMinusSuffix := strings.TrimSuffix(endReasonEnumsMinusPrefix, "}")
+						endReasons := strings.Split(endReasonEnumsMinusSuffix, ",")
+
+						for endReasonIndex, endReason := range endReasons {
+							if _, ok := knownEndReasons[endReason]; !ok {
+								errorsChan <- fmt.Errorf("Unknown end reason '%s' found in end_reason enum. Please report.", endReason)
+							}
+							endReasonMap[endReasonIndex] = endReason
+						}
+					}
+				}
+			}
+
+			// Get the read headers and their identifiers. Though the primary read headers are in a defined order, the auxiliary headers are not.
+			if values[0] == "#read_id" {
+				headerMap[0] = "read_id"
+				for headerNum := 1; headerNum < len(values); headerNum++ {
+					headerMap[headerNum] = values[headerNum]
+				}
+				start = true
+				continue
+			}
+		}
+
+		// Once we have the read headers, start to parse the actual reads
+		if start {
```

Comment:
I think this line can be removed so that all the block that comes after it is unindented and will be always called when `!start`, then you add a `continue` after line 183 just to ensure the loop continues, so:

```go
if !start {
  ...
  continue
}

// Once we have the read headers, start to parse the actual reads
var newRead Read
for valueIndex := 0; valueIndex < len(values); valueIndex++ {
...
```